providers/implementations/kdfs/snmpkdf.c

```diff
@@ -0,0 +1,314 @@
+}
+
+const OSSL_DISPATCH ossl_kdf_snmpkdf_functions[] = {
+    { OSSL_FUNC_KDF_NEWCTX, (void(*)(void))kdf_snmpkdf_new },
+    { OSSL_FUNC_KDF_DUPCTX, (void(*)(void))kdf_snmpkdf_dup },
+    { OSSL_FUNC_KDF_FREECTX, (void(*)(void))kdf_snmpkdf_free },
+    { OSSL_FUNC_KDF_RESET, (void(*)(void))kdf_snmpkdf_reset },
+    { OSSL_FUNC_KDF_DERIVE, (void(*)(void))kdf_snmpkdf_derive },
+    { OSSL_FUNC_KDF_SETTABLE_CTX_PARAMS,
+      (void(*)(void))kdf_snmpkdf_settable_ctx_params },
+    { OSSL_FUNC_KDF_SET_CTX_PARAMS, (void(*)(void))kdf_snmpkdf_set_ctx_params },
+    { OSSL_FUNC_KDF_GETTABLE_CTX_PARAMS,
+      (void(*)(void))kdf_snmpkdf_gettable_ctx_params },
+    { OSSL_FUNC_KDF_GET_CTX_PARAMS, (void(*)(void))kdf_snmpkdf_get_ctx_params },
+    OSSL_DISPATCH_END
+};
+
+
+
+
+/*
+  SNMPKDF - In compliance with SP800-135 and RFC 7860, calculate
+             a master key using the engine ID and password.
+
+  Denote engineLength and passwordlen to be the lengths (in bytes) of an
+  snmpEngineID and a password, respectively.
+
+  Let N = (1024*1024)/passwordlen
+
+  Expanded_password = the leftmost 1048576 bytes of the string of N
+  repetitions of the password.
+
+  Derived_password = SHA-1 (Expanded_password). The Derived_password
+  is the output of hashing the Expanded_password by SHA-1.
+
+  Let Shared_key to be the key that the user shares with the authoritative
+  SNMP engine with ID snmpEngineID. The Shared_key is generated as follows:
+
+  Shared_key = SHA-1(Derived_password || snmpEngineID || Derived_password).
+
+      e_id -         engine ID(eid)
+      e_len -        engineID length
+      password -     password
+      password_len - password length
+      okey -         pointer to key output, FIPS testing limited to SHA-1.
+      return -       1 pass 0 for error
+ */
+static int SNMPKDF(const EVP_MD *evp_md,
+                   const unsigned char *e_id, size_t e_len,
+                   unsigned char *password, size_t password_len,
+                   unsigned char *okey, size_t okeylen)
+{
+    EVP_MD_CTX *md = NULL;
+    unsigned char digest[EVP_MAX_MD_SIZE];
+    size_t mdsize = 0, len = 0;
+    unsigned int md_len = 0;
+    int ret = 0;
+
+    /* Limited to SHA-1 and SHA-2 hashes presently */
+    if (okey == NULL || okeylen == 0)
+        return 0;
+
+    md = EVP_MD_CTX_new();
+    if (md == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MESSAGE_DIGEST);
+        goto err;
+    }
+
+    mdsize = EVP_MD_size(evp_md);
+    if (mdsize <= 0 || mdsize < okeylen)
+        goto err;
+
+    if (!EVP_DigestInit_ex(md, evp_md, NULL))
+        goto err;
+
+    for (len = 0; len < KDF_SNMP_PASSWORD_HASH_AMOUNT - password_len; len += password_len ) {
+        if (!EVP_DigestUpdate(md, password, password_len)) {
+            goto err;
+        }
+    }
+
+    if (!EVP_DigestUpdate(md, password, KDF_SNMP_PASSWORD_HASH_AMOUNT - len))
+        goto err;
+
+    if (!EVP_DigestFinal_ex(md, digest, &md_len))
+        goto err;
```

Comment:
Normally this code would be collapsed..

````
if (!EVP_DigestInit_ex(md, evp_md, NULL)
        || !
        || !  )
    goto err;
 ````

Reply:
done